coins/src/utils/coins3/produce.ts

```diff
@@ -15,102 +18,170 @@ export type Dynamo = {
-    decimals,
-    price,
-    confidence,
-    SK,
-    adapter,
-    redirect,
-    mcap,
-  } = item;
-
-  const { chain, address, pid } = splitPk(PK);
-  const redirectPid = redirect ? splitPk(redirect).pid : undefined;
+  const { PK, decimals, price, confidence, SK, adapter, mcap } = item;
+  const { pid } = splitPk(PK, decimals);
 
   switch (topic) {
     case "coins-metadata":
-      return {
-        symbol,
-        decimals: Number(decimals),
-        address,
-        pid,
-        chain,
-        source: adapter,
-        redirect: redirectPid,
-      };
+      return; // for coins-metadata, we now rely on getMetadataRecord elsewhere
     case "coins-current":
-      return { pid, price, confidence, source: adapter, mcap };
+      return { pid, price, confidence, adapter, mcap, updateTs: Math.floor(Date.now() / 1000) };
     case "coins-timeseries":
-      return { pid, price, confidence, source: adapter, ts: SK, mcap };
+      return { pid, price, confidence, adapter, ts: SK, mcap };
     default:
       throw new Error(`Topic '${topic}' is not valid`);
   }
 }
-function splitPk(pk: string): { chain: string; address: string; pid: string } {
-  const assetPrefix: string = "asset#";
-  const coingeckoPrefix: string = "coingecko#";
-
-  if (pk.toLowerCase().startsWith(coingeckoPrefix)) {
-    const address = pk.substring(coingeckoPrefix.length).toLowerCase();
-    return {
-      chain: "coingecko",
-      address,
-      pid: `coingecko:${address}`,
-    };
+
+function splitPk(pk: string, decimals?: number): { chain?: string; address?: string; pid: string } {
+  const pid = normalizeCoinId(pk);
+  const record: any = { pid };
+  if (pid.includes(":")) {
+    record.chain = pid.split(":")[0];
+    record.address = pid.slice(record.chain.length + 1);
+  } else if (pid.length === 42 && pid.startsWith("0x")) {
+    record.chain = "ethereum";
+    record.address = pid;
+  } else if (!decimals && !pid.startsWith("0x")) {
+    record.chain = "coingecko";
+  }
+  return record;
+}
+
+async function produceMetadata(items: Dynamo[], producer: Producer) {
+  try {
+    await init();
+  } catch (initError) {
+    console.error("Error initializing coin metadata cache:", initError);
+  }
+
+  const metadataMessages: string[] = [];
+
+  items.map((item) => {
+    const record = getMetadataRecord(item);
+    if (!record) return;
+    validate(record, "coins-metadata");
+
+    const msg = JSON.stringify(record);
+    if (!metadataMessages.includes(msg)) {
+      metadataMessages.push(msg);
+    }
+  });
+
+  if (process.env.KAFKA_UP === "true") {
+    try {
+      await producer.send({
+        topic: "coins-metadata",
+        messages: metadataMessages.map((value) => ({ value })),
+      });
+      return;
+    } catch (error) {
+      console.error('Kafka failed for "coins-metadata". Fallback to ES.', error);
+    }
+  }
+
+  try {
+    const body: any[] = [];
+    for (const msg of metadataMessages) {
+      const doc = JSON.parse(msg);
```

Comment:
need to set document `_id` here as pid, else ES autogenerates it and we end up with duplicates